src/lib/spinel/radio_spinel_impl.hpp

```diff
@@ -1696,5 +1724,71 @@ otRadioState RadioSpinel<InterfaceType, ProcessContextType>::GetState(void) cons
     return sOtRadioStateMap[mState];
 }
 
+template <typename InterfaceType, typename ProcessContextType>
+void RadioSpinel<InterfaceType, ProcessContextType>::CalcRcpTimeOffset(void)
+{
+#if OPENTHREAD_CONFIG_MAC_CSL_RECEIVER_ENABLE || OPENTHREAD_CONFIG_MAC_CSL_TRANSMITTER_ENABLE
+    otError        error = OT_ERROR_NONE;
+    uint64_t       localTxTimestamp;
+    uint64_t       localRxTimestamp;
+    uint64_t       remoteTimestamp = 0;
+    uint8_t        buffer[sizeof(remoteTimestamp)];
+    spinel_ssize_t packed;
+
+    otLogInfoPlat("Trying to get RCP time offset");
+
+    /*
+     * Use a modified Network Time Protocol(NTP) to calculate the time offset
+     * Assume the time offset is D so that local can calculate remote time with,
+     *         T' = T + D
+     * Where T is the local time and T' is the remote time.
+     * The time offset is calculated using timestamp measured at local and remote.
+     *
+     *              T0  P    P T2
+     *  local time --+----+----+--->
+     *                \   |   ^
+     *              get\  |  /is
+     *                  v | /
+     * remote time -------+--------->
+     *                    T1'
+     *
+     * Based on the assumptions,
+     * 1. If the propagation time(P) from local to remote and from remote to local are same.
+     * 2. Both the host and RCP can accurately measure the time they send or receive a message.
+     * The degree to which these assumptions hold true determines the accuracy of the offset.
+     * Then,
+     *         T1' = T0 + P + D and T1' = T2 - P + D
+     * Time offset can be calculated with,
+     *         D = T1' - ((T0 + T2)/ 2)
+     */
+
+    VerifyOrExit(otPlatTimeGet() >= GetNextRadioTimeRecalcStart(), OT_NOOP);
+    packed = spinel_datatype_pack(buffer, sizeof(buffer), SPINEL_DATATYPE_UINT64_S, remoteTimestamp);
+    VerifyOrExit(packed > 0 && static_cast<size_t>(packed) <= sizeof(buffer), error = OT_ERROR_NO_BUFS);
+
+    localTxTimestamp = otPlatTimeGet();
+
+    // Dummy timestamp payload to make request length same as response
+    error = GetWithParam(SPINEL_PROP_RCP_TIMESTAMP, buffer, packed, SPINEL_DATATYPE_UINT64_S, &remoteTimestamp);
+
+    localRxTimestamp = otPlatTimeGet();
+
+    VerifyOrExit(error == OT_ERROR_NONE, mRadioTimeRecalcStart = localRxTimestamp);
+
+    mRadioTimeOffset      = remoteTimestamp - ((localRxTimestamp / 2) + (localTxTimestamp / 2));
+    mIsTimeSynced         = true;
+    mRadioTimeRecalcStart = localRxTimestamp + RCP_TIME_OFFSET_CHECK_INTERVAL;
+
+exit:
+    LogIfFail("Error calculating RCP time offset: %s", error);
+#endif
+}
+
+template <typename InterfaceType, typename ProcessContextType>
+uint64_t RadioSpinel<InterfaceType, ProcessContextType>::GetNow(void)
+{
+    return mIsTimeSynced ? (otPlatTimeGet() + static_cast<uint64_t>(mRadioTimeOffset)) : 0;
```

Comment:
`0` is a valid time, how about using UINT64_MAX
```suggestion
    return mIsTimeSynced ? (otPlatTimeGet() + static_cast<uint64_t>(mRadioTimeOffset)) : UINT64_MAX;
```

Reply:
Good idea!